habitat/core/dataset.py

```diff
@@ -236,20 +243,116 @@ def get_splits(
+        Cycling: when all episodes are iterated, cycle back to start instead of
+            throwing StopIteration.
+        Cycling with shuffle: when cycling back, shuffle episodes groups
+            grouped by scene.
+        Group by scene: episodes of same scene will be grouped and loaded
+            consecutively.
+        Set max scene repeat: set a number threshold on how many episodes from
+        the same scene can be loaded consecutively.
+        Sample episodes: sample the specified number of episodes.
+    """
+
+    def __init__(
+        self,
+        episodes: List[T],
+        cycle: bool = True,
+        shuffle_scene_when_cycle: bool = False,
+        group_by_scene: bool = True,
+        max_scene_repeat: int = -1,
+        num_episode_sample: int = -1,
+    ):
+        r"""
         Args:
-            num_episodes: number of episodes to sample, input -1 to use
-            whole episodes
+            episodes: list of episodes.
+            cycle: if true, cycle back to first episodes when StopIteration.
+            shuffle_scene_when_cycle: if true, shuffle scene groups when cycle.
+                No effect if cycle is set to false.
+            group_by_scene: if true, group episodes from same scene.
+            max_scene_repeat: threshold of how many episodes from the same
+                scene can be loaded consecutively. -1 for no limit
+            num_episode_sample: number of episodes to be sampled.
+                -1 for no sampling.
         """
-        if num_episodes == -1:
-            return
-        if num_episodes < -1:
-            raise ValueError(
-                f"Invalid number for episodes to sample: {num_episodes}"
+        # sample episodes
+        if num_episode_sample >= 0:
+            episodes = np.random.choice(
+                episodes, num_episode_sample, replace=False
             )
-        self.episodes = np.random.choice(
-            self.episodes, num_episodes, replace=False
-        )
+        self.episodes = episodes
+        self.cycle = cycle
+        self.group_by_scene = group_by_scene
+        if group_by_scene:
+            num_scene_groups = len(
+                list(groupby(episodes, key=lambda x: x.scene_id))
+            )
+            num_unique_scenes = len(set([e.scene_id for e in episodes]))
+            if num_scene_groups >= num_unique_scenes:
+                self.episodes = sorted(self.episodes, key=lambda x: x.scene_id)
+        self.max_scene_repetition = max_scene_repeat
+        self.shuffle_scene_when_cycle = shuffle_scene_when_cycle
+        self._rep_count = 0
+        self._prev_scene_id = None
+        self._iterator = iter(self.episodes)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        r"""The main logic for handling how episodes will be iterated.
+
+        Returns:
+            next episode.
+        """
+        try:
+            next_episode = next(self._iterator)
```

Comment:
Just double checking, is there any other better way to handle this besides catching an exception, if not this seems good.

Reply:
As far as I know, there's no simple way to check if the internal iterator has exhausted all episodes.

Reply:
You can do
```python
next_episode = next(self._iterator, None)
if next_episode is None:
    ....
```